subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,264 @@
+package targets
+
+import (
+	"archive/tar"
+	"compress/bzip2"
+	"fmt"
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"io"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	factory string
+	prod    bool
+	tag     string
+)
+
+func init() {
+	offlineUpdateCmd := &cobra.Command{
+		Use:   "offline-update <target-name> <dst>",
+		Short: "Download Target content for an offline update",
+		Run:   doOfflineUpdate,
+		Args:  cobra.ExactArgs(2),
+	}
+	cmd.AddCommand(offlineUpdateCmd)
+	offlineUpdateCmd.Flags().BoolVarP(&prod, "prod", "", false,
+		"Instruct to fetch content of production Target")
+	offlineUpdateCmd.Flags().StringVarP(&tag, "tag", "", "master",
+		"Target tag")
+}
+
+func doOfflineUpdate(cmd *cobra.Command, args []string) {
+	factory = viper.GetString("factory")
+	targetName := args[0]
+	dstDir := args[1]
+
+	fmt.Printf("Downloading offline update content of Target %s to %s\n", targetName, dstDir)
+
+	fmt.Printf("Getting info about Target %s \n", targetName)
+
+	// We get the Target's custom data from the list of all targets regardless whether it's a production target or not
+	// because we need to obtain the Target's tag that was used to build OE and Apps/containers, which in turn is needed
+	// to deduce a path to the Apps archive artifact.
+	// E.g. https://ci.foundries.io/projects/msul-dev01/lmp/builds/1420/assemble-system-image/ is built as `master`
+	// while in the production targets the corresponding target is marked as `prod`.
+	// TODO: switch to usage of /api/v1/user_repo/targets.json API once it supports `production` param
+	// instead of doing the stuff below
+	custom, err := getTargetCustomInfo(targetName)
```

Comment:
I'm not sure how is this supposed to work, especially for the prod targets (which require both online and an offline signatures).

I believe that for offline update we should mimic what the device does during an online update as close as possible in order to be fully interoperable and respect the TUF guidelines. Respective this line I think we should fetch targets.json by tag (just like the device does) and search for all necessary information inside that specific tag-based targets.json file, not the CI targets.json. The device never downloads the CI targets.json, and in fact data in it may be different from the tag-based targets.json, especially when we take production targets into consideration.

We have an API to get tag-based TUF roles already (your new `TufMetadataGet` method does it right), just need to extend it with the production targets support. An interesting use case would be to extend it with the wave targets support (by wave name or device group name???).

Reply:
> Respective this line I think we should fetch targets.json by tag (just like the device does) 

This line does NOT fetch targets.json for an offline update, the downloaded targets as well as the other role metadata are fetched via the API you suggest (api/v1/user_repo/).

> getTargetCustomInfo is needed solely to get a tag (effectively branch) that was used at a CI build. The corresponding prod target has different tag than the tag that was used to build the given target. Therefore, we cannot use prod targets for it, only CI one.

I don't like it too, so I think we should change the way we store the apps artifact in a CI run so it is tag agnostic...

Reply:
> > Respective this line I think we should fetch targets.json by tag (just like the device does)
> 
> This line does NOT fetch targets.json for an offline update, the downloaded targets as well as the other role metadata are fetched via the API you suggest (api/v1/user_repo/).
> 
> > getTargetCustomInfo is needed solely to get a tag (effectively branch) that was used at a CI build. The corresponding prod target has different tag than the tag that was used to build the given target. Therefore, we cannot use prod targets for it, only CI one.
> 
> I don't like it too, so I think we should change the way we store the apps artifact in a CI run so it is tag agnostic...

I think I understand the use case, and it troubles me even more this way, as this code ties the user to our CI artifacts storage. I agree it may be a convenient way to simplify the implementation. However, this looks like a completely different way to fetch targets artifacts compared to the way device does that.

Is it possible to fetch those artifacts from the same storage the device does, e.g. implement the ostree-proxy like signed-url handler in ota-lite or support the user auth in ostree-proxy (like we do for e.g. hub.fio)?

I suspect that could turn into a "feature" on itself, and we might have zero customers for this. That said, you can ignore this comment, I'm just trying to think of 100500 use cases which could be completely irrelevant to you work.

Reply:
Yep, there are many aspects, it's better to discuss it offline.